### ISSUE TITLE
sql support

### ISSUE THREAD

**Issue by mg1075** [meanjs/mean#1161]
Much of the goodness in MEAN.JS exists in the way it brings structure to node applications and how it tries to bring in various node pieces to address common concerns for building web applications. The 0.5.0 roadmap appears to address even more of those concerns: e.g., using JWTs.

The M in MEAN.JS is troublesome, though. MongoDB is **not** the database of choice in many a corporation, nor is it the go-to database for relational data, even though Mongoose - which is the true M in MEAN.JS - tries to make it seem like you are working with a relational db. Therefore, **please consider adding sql support** (at least PostgreSQL, MySQL, and SQLite3).

The project would likely gain lots more traction if equal importance was given to sql support. This would probably have to end up being a parallel/sister project: it would be like MEAN.JS in all possible ways except for the MongoDB/Mongoose dependency. Even though there are some github projects that sort of try to do this with MEAN, none of them have any traction or ongoing maintenance and development.

Maybe it would use knex, or bookshelf, or sequelize, or whatever query builder/orm tool for sql support - that is a separate discussion. 

I guess you could make the same argument about the A in MEAN and say you really should instead be using framework [whatever] instead of Angular, and if we go that route, the concept of the MEAN.JS kind of falls apart. Fair enough. You can't have a framework/boilerplate that is going to make everyone happy!

- for reference, the only related issue I read on this topic is here:
https://github.com/meanjs/mean/issues/351

**Comment by ilanbiala** [comment]
@codydaig @lirantal @mleanos I'd like to hear your suggestions to see if I might not be considering something, but I think this is too much of a shift and we won't be able to support both platforms.

**Comment by mg1075** [comment]
Seriously...just found out about this project this afternoon on Reddit after scouring the interwebs. The project seems to do pretty much exactly what I hoped for (except not using Knex):

PEAN.JS
https://github.com/StetSolutions/pean

- same as MEAN 0.4.2
- in place of Mongoose, the ORM is Sequelize
- possibly more traction than the other abandoned projects out there

**Maybe you all should collaborate so both projects stay in sync as much as possible?**

**Comment by codydaig** [comment]
I'm open to collaboration. However, I'd agree that it's a huge shift and not the purpose of this project. Glancing at PEAN, I like how it is using this code base, and since it maintains the same history, it should allow for changes to be pulled in if someone wants to put the effort in.

**Comment by mleanos** [comment]
IMO, the simplicity of MongoDB (& mongoose) is great for a framework like this. If someone **must** use a relational database, then they can implement one into their project. If we start off with having support for other ORM's (Relational db's) then the project will suffer because it would then become more opinionated.

With that said, I think we can find ways to make it easier to switch to a new database technology, or module. Currently, the way the project is setup, we're pretty tied to Mongoose. As with a framework like this, I think the simplicity, and non-opinionated implementation, is what makes it attractive. And it's meant to be built upon.

**Comment by lirantal** [comment]
@ilanbiala I think it's not such a bad idea to try and add some MySQL or PostgreSQL connectors in the schema/project. We don't need to actually enforce and use it in the project but it could be a good start for anyone whose thinking of using MEAN.JS to also just connect with a "regular" database server as an API component and such.

**Comment by anterokarttunen** [comment]
TLDR, but would be cool to be able to choose whether to use mongoose or sequelize (or both :dancer: ) - then just have some if clauses, if(mongooseEnabled) or if(sequelizeEnabled)

**Comment by lirantal** [comment]
@anterokarttunen this post is actually about supporting MySQL.

**Comment by anterokarttunen** [comment]
@lirantal Ok - would that mean also mariadb is supported? since its 100% drop in replacement for mysql.

Are you planning to introduce some kind of "node sql framework" or build sql support from scratch ?

**Comment by lirantal** [comment]
MariaDB and MySQL are indeed very similar.
There's no plan though on adding any special SQL framework support yet.

@meanjs/contributors WDYT about MySQL support?

**Comment by mg1075** [comment]
To clarify, my original post concerned supporting SQL in a broader sense than just MySQL.

> Therefore, please consider adding sql support (at least PostgreSQL, MySQL, and SQLite3).

To date, the framework that most closely does this, forked off of MEAN, is PEAN, which makes use of sequelize as its orm. 

https://github.com/StetSolutions/pean

On a related note, this article can at least help others appreciate why mongodb/mongoose is often not going to be the first choice of db's in a business environment.
http://www.bitnative.com/2015/04/13/why-im-not-sold-on-mongodb/

**Comment by codydaig** [comment]
I'm not opposed to making the framework in a way that would support ANY database type really. My thoughts here would be to take the DB logic out of the controllers and into a 'models' (or whatever we want to call it) file. 

ie. You would call Users.create() which would then take that data and interact with the database. Then all one would have to do to swap databases would be change the db connection point and those model files. 

Personally, I don't want to support multiple DBs directly in the framework, but I'm open to making the DB portion more modular and easily swappable.

**Comment by lirantal** [comment]
My concern is that I also don't want to tightly couple MEAN.JS with other libraries like sequelize too much. The best part I love about MEAN.JS is that it's very flexible and even if you do MySQL and not MongoDB it is very easy to replace that logic in the framework and do your stuff.


@mg1075 on another note, I heard that Waterfall for example is an amazing performance hog. Have you got any statistics on how good sequalize is? which other frameworks have adopted it?

**Comment by ilanbiala** [comment]
I think the only logical way for us to support multiple databases in the frameworks would be to remove database support and write stubbed methods that users would have to fill in. That solution obviously have disadvantages for everyone, and honestly writing in more code to support 2 database types is also with its disadvantages.

**Comment by lirantal** [comment]
Agree.
I'm not yet sold as to which path to take and doesn't seem to be any traction on this topic from what I've been reading on the issue queue so far.

**Comment by Wuntenn** [comment]
When you look at the reasons for using PEAN they all seem valid. The argument that most apps aren't document based and that really they're relational and that all the big companies that have used MEAN have had to convert seems like a strong argument.

However, a lot of MEAN applications function as small side-projects and/or early startup code base where there is constant change. In the beginning of a project where there are the highest number of unknowns it's probably better to have the flexibility that MEAN offers without the headache of learning SQL writing (and supporting) migrations.

It's always possible to convert more established slower changing Schema to SQL later where your priority shifts to maximising maintenance/performance/reliability.

Done this way around meanjs projects start at their most accessible state with the ability to transition to more robust solutions. This is a huge strength to have in my opinion and whilst this project and its users may benefit from supporting PEAN I'm also not sold on switching.

**Comment by lirantal** [comment]
Thanks for the discussion everyone.
